tools/src/main/java/org/apache/kafka/trogdor/workload/ExternalCommandWorker.java

```diff
@@ -116,43 +115,57 @@ public void start(Platform platform, WorkerStatusTracker status,
     public class ProcessMonitor implements Runnable {
         @Override
         public void run() {
+            if (spec.command().isEmpty()) {
+                errMsg = "No command specified";
+                doneFuture.complete(errMsg);
+                return;
+            }
+
             try {
-                if (spec.command().isEmpty()) {
-                    errMsg = "No command specified";
-                    doneFuture.complete(errMsg);
-                    return;
-                }
-                ProcessBuilder processBuilder = new ProcessBuilder(spec.command());
-                process = processBuilder.start();
-                Future updater = executor.submit(new StatusUpdater());
-                Future errorUpdater = executor.submit(new ErrorUpdater());
-                controlChannel = new PrintWriter(new OutputStreamWriter(process.getOutputStream(), StandardCharsets.UTF_8), true);
-                new ControlCommand("start", spec).issue();
-                log.info("{}: Started the process {} to execute command: {}", id, process.toString(), spec.command());
-                process.waitFor();
-                updater.get();
-                errorUpdater.get();
-                int workerExitValue = process.exitValue();
+                ep = new ExternalProcess();
+            } catch (IOException e) {
+                errMsg = "Failed to start the external process.";
+                log.error("{}: Failed to start the external process.");
+                doneFuture.complete(errMsg);
+                return;
+            }
+
+            try {
+                ep.startTask();
```

Comment:
I think a race condition exist where `destroyProcess()` can be called right before this. In destroyProcess, `stopTask()` can get called (sending a STOP message to the process) and then `ep.startTask()` can get called, sending a START JSON message to the process.

While the process should probably be made in a way to ensure it stops once it receives a message, should we make sure we don't have the ability to send such mixed signals?

I'm not exactly sure what the simplest fix here would be. First of all, I think it makes sense to make every method in `ExternalProcess` `synchronized`. Then, we could maybe set an `boolean isClosed` in `destroyProcess()` and the other methods can always check for that boolean before doing anything
Thoughts @cmccabe @yangxi ?